diff --git a/sysdeps/posix/getaddrinfo.c b/sysdeps/posix/getaddrinfo.c
index ed04e564f9..b199570227 100644
--- a/sysdeps/posix/getaddrinfo.c
+++ b/sysdeps/posix/getaddrinfo.c
@@ -1071,6 +1071,8 @@ gaih_inet (const char *name, const struct gaih_service *service,
 	       terminated.  */
 	    ai->ai_next = NULL;
 
+# pragma GCC diagnostic push
+# pragma GCC diagnostic ignored "-Wstringop-overflow"
 	    if (family == AF_INET6)
 	      {
 		struct sockaddr_in6 *sin6p =
@@ -1091,6 +1093,7 @@ gaih_inet (const char *name, const struct gaih_service *service,
 			at2->addr, sizeof (struct in_addr));
 		memset (sinp->sin_zero, '\0', sizeof (sinp->sin_zero));
 	      }
+# pragma GCC diagnostic pop
 
 	    pai = &(ai->ai_next);
 	  }
