diff --git a/math/Makefile b/math/Makefile
index 84a8b94c74..a3ae648548 100644
--- a/math/Makefile
+++ b/math/Makefile
@@ -659,6 +659,10 @@ CFLAGS-s_modff.c += -fsignaling-nans
 CFLAGS-s_modfl.c += -fsignaling-nans
 CFLAGS-s_modff128.c += -fsignaling-nans
 
+# Work around GCC diagnostic
+CFLAGS-s_modf.c += -Wno-error=builtin-declaration-mismatch
+CFLAGS-s_sincos.c += -Wno-error=builtin-declaration-mismatch
+
 $(addprefix $(objpfx),\
 	    $(filter-out $(tests-static) $(libm-tests-vector),\
 			 $(tests) $(tests-internal))): $(libm)
