diff --git a/elf/Makefile b/elf/Makefile
index 77da61ae64..15f84ecd72 100644
--- a/elf/Makefile
+++ b/elf/Makefile
@@ -51,6 +51,9 @@ CFLAGS-dl-tunables.c += -fno-tree-loop-distribute-patterns
 endif
 endif
 
+# Hack until either GCC or glibc is fixed
+CFLAGS-dl-support.c += -Wno-uninitialized
+
 all-dl-routines = $(dl-routines) $(sysdep-dl-routines)
 # But they are absent from the shared libc, because that code is in ld.so.
 elide-routines.os = $(all-dl-routines) dl-support enbl-secure dl-origin \
