diff --git a/iconv/loop.c b/iconv/loop.c
index 062cc1b868..3ce8892ca7 100644
--- a/iconv/loop.c
+++ b/iconv/loop.c
@@ -436,6 +436,8 @@ SINGLE(LOOPFCT) (struct __gconv_step *step,
     return __GCONV_FULL_OUTPUT;
 
   /*  Now add characters from the normal input buffer.  */
+  if (inlen >= MAX_NEEDED_INPUT)
+    __builtin_unreachable ();
   do
     bytebuf[inlen++] = *inptr++;
   while (inlen < MAX_NEEDED_INPUT && inptr < inend);
