diff --git a/gcc/testsuite/gcc.c-torture/execute/pr60822.c b/gcc/testsuite/gcc.c-torture/execute/pr60822.c
index dcd2447..b326e17 100644
--- a/gcc/testsuite/gcc.c-torture/execute/pr60822.c
+++ b/gcc/testsuite/gcc.c-torture/execute/pr60822.c
@@ -1,4 +1,5 @@
 /* { dg-require-effective-target int32plus } */
+/* { dg-do run { target { ! simulator } } } */
 struct X {
     char fill0[800000];
     int a;
diff --git a/gcc/testsuite/gcc.dg/torture/ftrapv-1.c b/gcc/testsuite/gcc.dg/torture/ftrapv-1.c
index 66fa33f..9385c19 100644
--- a/gcc/testsuite/gcc.dg/torture/ftrapv-1.c
+++ b/gcc/testsuite/gcc.dg/torture/ftrapv-1.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-additional-options "-ftrapv -fno-ipa-vrp" } */
 /* { dg-require-effective-target trapping } */
 /* { dg-require-fork "" } */
diff --git a/gcc/testsuite/gcc.dg/torture/ftrapv-2.c b/gcc/testsuite/gcc.dg/torture/ftrapv-2.c
index 8065ee0..51af15f 100644
--- a/gcc/testsuite/gcc.dg/torture/ftrapv-2.c
+++ b/gcc/testsuite/gcc.dg/torture/ftrapv-2.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* With -flto this degenerates to constant folding which doesn't work.  */
 /* { dg-skip-if "" { *-*-* } { "-flto" } { "" } } */
 /* { dg-additional-options "-ftrapv" } */
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1.c
index 4a81e12..0acf1d4 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange -fassociative-math -fno-signed-zeros -fno-trapping-math -fdump-tree-linterchange-details" } */
 /* { dg-skip-if "too big data segment" { avr-*-* visium-*-* } } */
 
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-15.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-15.c
index fca70d6..48a1e49 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-15.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-15.c
@@ -1,9 +1,10 @@
 /* PR tree-optimization/83337 */
-/* { dg-do run { target int32plus } } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange" } */
 /* { dg-require-effective-target alloca }  */
 /* { dg-skip-if "too big stack" { visium-*-* } } */
 
+
 /* Copied from graphite/interchange-5.c */
 
 #define DEBUG 0
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1b.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1b.c
index 7de2b40..aad0a0e 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1b.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-1b.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange -fdump-tree-linterchange-details" } */
 /* { dg-skip-if "too big data segment" { avr-*-* visium-*-* } } */
 
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-5.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-5.c
index 8f13cd4..cc9adf5 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-5.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-5.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange -fdump-tree-linterchange-details" } */
 /* { dg-skip-if "too big data segment" { avr-*-* visium-*-* } } */
 
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-6.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-6.c
index a38fd9d..8a65223 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-6.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-6.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange -fdump-tree-linterchange-details" } */
 /* { dg-skip-if "too big data segment" { avr-*-* visium-*-* } } */
 
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-7.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-7.c
index e6ad26f..f487bef 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-7.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-7.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange -fdump-tree-linterchange-details" } */
 /* { dg-skip-if "too big data segment" { avr-*-* visium-*-* } } */
 
diff --git a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-8.c b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-8.c
index 5ed300a..9c7830a 100644
--- a/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-8.c
+++ b/gcc/testsuite/gcc.dg/tree-ssa/loop-interchange-8.c
@@ -1,4 +1,4 @@
-/* { dg-do run } */
+/* { dg-do run { target { ! simulator } } } */
 /* { dg-options "-O2 -floop-interchange -fdump-tree-linterchange-details" } */
 /* { dg-skip-if "too big data segment" { avr-*-* visium-*-* } } */
 
