diff --git a/gcc/haifa-sched.c b/gcc/haifa-sched.c
index 1d3de7b6a76..80687fb5359 100644
--- a/gcc/haifa-sched.c
+++ b/gcc/haifa-sched.c
@@ -4239,6 +4239,15 @@ remove_notes (rtx_insn *head, rtx_insn *tail)
 	  if (insn != tail)
 	    {
 	      remove_insn (insn);
+	      /* If an insn was split just before the EPILOGUE_BEG note and
+		 that split created new basic blocks, we could have a
+		 BASIC_BLOCK note here.  Safely advance over it in that case
+		 and assert that we land on a real insn.  */
+	      if (NOTE_P (next)
+		  && NOTE_KIND (next) == NOTE_INSN_BASIC_BLOCK
+		  && next != next_tail)
+		next = NEXT_INSN (next);
+	      gcc_assert (INSN_P (next));
 	      add_reg_note (next, REG_SAVE_NOTE,
 			    GEN_INT (NOTE_INSN_EPILOGUE_BEG));
 	      break;
