diff --git a/gcc/config/h8300/h8300.md b/gcc/config/h8300/h8300.md
index bcc78a4ce4d..fdd2d8b02d7 100644
--- a/gcc/config/h8300/h8300.md
+++ b/gcc/config/h8300/h8300.md
@@ -4299,6 +4299,7 @@
 	(plus:SI (match_dup 0)
 		 (match_operand:SI 2 "const_int_operand" "")))]
   "(TARGET_H8300H || TARGET_H8300S)
+    && operands[0] != stack_pointer_rtx
     && REG_P (operands[0]) && REG_P (operands[1])
     && REGNO (operands[0]) != REGNO (operands[1])
     && !satisfies_constraint_L (operands[2])
diff --git a/gcc/cse.c b/gcc/cse.c
index 3e8724b3fed..d1efc0afc9c 100644
--- a/gcc/cse.c
+++ b/gcc/cse.c
@@ -4716,7 +4716,8 @@ cse_insn (rtx_insn *insn)
       /* Compute SRC's hash code, and also notice if it
 	 should not be recorded at all.  In that case,
 	 prevent any further processing of this assignment.  */
-      do_not_record = 0;
+      do_not_record = (REG_P (dest)
+		       ? find_reg_note (insn, REG_UNUSED, dest) != 0 : 0);
       hash_arg_in_memory = 0;
 
       sets[i].src = src;
