diff --git a/gcc/testsuite/gcc.dg/cpp/charconst-3.c b/gcc/testsuite/gcc.dg/cpp/charconst-3.c
index 7c196639131..591ac742f5e 100644
--- a/gcc/testsuite/gcc.dg/cpp/charconst-3.c
+++ b/gcc/testsuite/gcc.dg/cpp/charconst-3.c
@@ -1,7 +1,7 @@
 /* Copyright (C) 2001 Free Software Foundation, Inc.  */
 
 /* { dg-do run } */
-/* { dg-options -Wno-multichar } */
+/* { dg-options -Wno-multichar -fwrapv} */
 
 /* This tests values and signedness of multichar charconsts.
 
