patchc3.txt

diff --git a/gcc/match.pd b/gcc/match.pd
index 4d41b70..f4c0c03 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -3644,8 +3644,9 @@ DEFINE_INT_AND_FLOAT_ROUND_FN (RINT)
    (cmp (bswap @0) (bswap @1))
    (cmp @0 @1))
   (simplify
-   (cmp (bswap @0) INTEGER_CST@1)
-   (cmp @0 (bswap @1))))
+   (cmp (bswap@0 @1) INTEGER_CST@2)
+   (with { tree t = TREE_TYPE(@0); }
+    (cmp (convert:t @1) (convert:t (bswap @2))))))
  /* (bswap(x) >> C1) & C2 can sometimes be simplified to (x >> C3) & C2.  */
  (simplify
   (bit_and (convert1? (rshift@0 (convert2? (bswap@4 @1)) INTEGER_CST@2))

