diff --git a/gcc/builtins.c b/gcc/builtins.c
index 6f67eb7eeb5..a85af2024ef 100644
--- a/gcc/builtins.c
+++ b/gcc/builtins.c
@@ -6775,9 +6775,9 @@ try_store_by_multiple_pieces (rtx to, rtx len, unsigned int ctz_len,
 
       /* Adjust PTR, TO and REM.  Since TO's address is likely
 	 PTR+offset, we have to replace it.  */
-      emit_move_insn (ptr, XEXP (to, 0));
+      emit_move_insn (ptr, force_reg (GET_MODE (ptr), XEPX (to, 0)));
       to = replace_equiv_address (to, ptr);
-      emit_move_insn (rem, plus_constant (ptr_mode, rem, -blksize));
+      expand_inc (rem, GEN_INT (-blksize));
     }
 
   /* Iterate over power-of-two block sizes from the maximum length to
@@ -6811,9 +6811,9 @@ try_store_by_multiple_pieces (rtx to, rtx len, unsigned int ctz_len,
       /* Adjust REM and PTR, unless this is the last iteration.  */
       if (i != sctz_len)
 	{
-	  emit_move_insn (ptr, XEXP (to, 0));
+	  emit_move_insn (ptr, force_reg (GET_MODE (ptr), XEXP (to, 0)));
 	  to = replace_equiv_address (to, ptr);
-	  emit_move_insn (rem, plus_constant (ptr_mode, rem, -blksize));
+	  expand_inc (rem, GEN_INT (-blksize));
 	}
 
       if (label)
