diff --git a/bfd/bfd-in.h b/bfd/bfd-in.h
index 453ac48e26..cf1913e648 100644
--- a/bfd/bfd-in.h
+++ b/bfd/bfd-in.h
@@ -566,6 +566,7 @@ struct bfd_link_hash_entry;
 
 /* Return TRUE if the start of STR matches PREFIX, FALSE otherwise.  */
 
+#include <string.h>
 static inline bfd_boolean
 startswith (const char *str, const char *prefix)
 {
diff --git a/bfd/bfd-in2.h b/bfd/bfd-in2.h
index c8bf687ba2..37a21d70d9 100644
--- a/bfd/bfd-in2.h
+++ b/bfd/bfd-in2.h
@@ -572,7 +572,7 @@ struct bfd_link_info;
 struct bfd_link_hash_entry;
 
 /* Return TRUE if the start of STR matches PREFIX, FALSE otherwise.  */
-
+#include <string.h>
 static inline bfd_boolean
 startswith (const char *str, const char *prefix)
 {
