diff --git a/sim/common/gentmap.c b/sim/common/gentmap.c
index a606fad0e4f..e3af481dbaf 100644
--- a/sim/common/gentmap.c
+++ b/sim/common/gentmap.c
@@ -71,6 +71,7 @@ gen_targ_map_c (void)
   printf ("#include <errno.h>\n");
   printf ("#include <fcntl.h>\n");
   printf ("#include \"ansidecl.h\"\n");
+  printf ("#include <string.h>\n");
   printf ("#include \"gdb/callback.h\"\n");
   printf ("#include \"targ-vals.h\"\n");
   printf ("\n");
