diff --git a/sim/common/sim-events.c b/sim/common/sim-events.c
index 207c8674d16..8c43b1224c0 100644
--- a/sim/common/sim-events.c
+++ b/sim/common/sim-events.c
@@ -469,10 +469,8 @@ sim_events_schedule (SIM_DESC sd,
 		     sim_event_handler *handler,
 		     void *data)
 {
-  va_list dummy;
-  memset (&dummy, 0, sizeof dummy);
   return sim_events_schedule_vtracef (sd, delta_time, handler, data,
-				      NULL, dummy);
+				      NULL);
 }
 #endif
 
