diff --git a/sim/m32r/cpu.h b/sim/m32r/cpu.h
index 9f8a0d1fbb..b5ceda8a39 100644
--- a/sim/m32r/cpu.h
+++ b/sim/m32r/cpu.h
@@ -36,6 +36,11 @@ This file is part of the GNU simulators.
 typedef CGEN_INSN_INT CGEN_INSN_WORD;
 
 #include "cgen-engine.h"
+#if 0
+extern DI   m32rbf_h_accum_get_handler (SIM_CPU *);
+extern void m32rbf_h_accum_set_handler (SIM_CPU *, DI);
+#endif
+#include "m32r-sim.h"
 
 /* CPU state information.  */
 typedef struct {
